### PR TITLE
Task 23998197: add winml_lib_core into onnnxruntime.dll

### PR DESCRIPTION
**Pull request by walrusmcd**
This change *adds* winml_lib_core into the onnxruntime sharelib.

this is a staging step, where we will start use that "core" as the adapter for winml, and it will live inside the ORT dll.

step1 is to add it, but still keep it in both dll's (win.ai.ml.dll AND ort.dll) .    the code needs to be clean such that it can live inside ORT and not have any connections back to WINML.

it also needs to be clean that it follows the ORT system of doing things (errors, telemetry, etc.)

### CODE REVIEW

File: winml/api/Windows.AI.MachineLearning.idl [hunk 1]
```diff
@@ -360,6 +383,23 @@ namespace Windows.AI.MachineLearning
...
+                String Name,
+                String Description,
+                Boolean IsRequired,
+                TensorKind TensorKind,
+                Int64[] Shape,
+                Boolean HasUnsupportedImageMetadata
+            );
+            //! if this feature is an image but has unsupport image metadata (like a BitmapPixelFormat) 
+            //! you can still use the runtime but without image conversion support.
+            //! this settig will be 'true'
```
Comment (ryanlai2) [inline winml/api/Windows.AI.MachineLearning.idl]: spelling: "setting"

File: winml/lib/Api/ImageFeatureDescriptor.cpp [hunk 2]
```diff
@@ -32,6 +32,28 @@ ImageFeatureDescriptor::ImageFeatureDescriptor(
                                               color_space_gamma_(color_space_gamma) {
 }
 
+ImageFeatureDescriptor::ImageFeatureDescriptor(
+        hstring const& Name,
```
Comment (ryanlai2) [inline winml/lib/Api/ImageFeatureDescriptor.cpp]: nit: the naming convention of these parameters don't match what's on the other constructor of ImageFeatureDescriptor.

File: winml/lib/Api/ImageFeatureDescriptor.h [hunk 3]
```diff
@@ -94,3 +105,9 @@ struct ImageFeatureDescriptor : ImageFeatureDescriptorT<
   ImageColorSpaceGamma color_space_gamma_;
 };
 }  // namespace winrt::Windows::AI::MachineLearning::implementation
+
+namespace winrt::Windows::AI::MachineLearning::factory_implementation {
+    struct ImageFeatureDescriptor : ImageFeatureDescriptorT<ImageFeatureDescriptor, implementation::ImageFeatureDescriptor> {
```
Comment (ryanlai2) [inline winml/lib/Api/ImageFeatureDescriptor.h]: Can you help me understand what this is for?

File: winml/lib/Api/ImageFeatureDescriptor.cpp (hunk 2 above)
Comment (walrusmcd) [reply winml/lib/Api/ImageFeatureDescriptor.cpp]: right, these got copied from the auto gen code.   we should decide how we want to do coding conventions, but my rough idea is something that follows along lineage.   so if autogen code does something a certain way, we stick that that.  if ort has another way, we stick with that.  etc.etc..  based on lineage.

---
In reply to: [344926984](https://github.com/microsoft/onnxruntime/pull/2368#discussion_r344926984) [](ancestors = 344926984)

File: winml/lib/Api/ImageFeatureDescriptor.h (hunk 3 above)
Comment (walrusmcd) [reply winml/lib/Api/ImageFeatureDescriptor.h]: these factories allow objects to be created.  prior to this change, external people could not create objects.   this change allows the creation of these objects.   factories are things that create objects.

---
In reply to: [344928039](https://github.com/microsoft/onnxruntime/pull/2368#discussion_r344928039) [](ancestors = 344928039)

File: winml/lib/Api.Core/FeatureDescriptorFactory.cpp [hunk 4]
```diff
@@ -438,31 +440,29 @@ CreateImageFeatureDescriptor(
   auto alpha_mode = format_info.second;
 
   // color space gamma value
-  auto color_space_gamma_value = FetchMetadataValueOrNull(metadata, c_color_space_key);
-  auto color_space_gamma = CreateImageColorSpaceGamma(color_space_gamma_value);
+//  auto color_space_gamma_value = FetchMetadataValueOrNull(metadata, c_color_space_key);
```
Comment (ryanlai2) [inline winml/lib/Api.Core/FeatureDescriptorFactory.cpp]: Why hardcode the creation of ImageFeatureDescriptor to 0-255 Nominal Range and SRGB color space gamma?

File: winml/lib/Api/ImageFeatureDescriptor.h (hunk 3 above)
Comment (ryanlai2) [reply winml/lib/Api/ImageFeatureDescriptor.h]: Thanks! That makes sense

**Pull request merged**